client/client.go

```diff
@@ -106,6 +106,9 @@ type Info struct {
 	// configauth.ServerAuthenticator implementations tied to the receiver for
 	// this connection.
 	Auth AuthData
+
+	// Metadata is the request metadata from the client connecting to this connector.
+	Metadata map[string][]string
```

Comment:
Sorry that I missed this. But because of map/slice usage here, this makes Info mutable which needs to be thread-safe and will complicate lots of things.

@jpkrohling we discussed about this and initially we thought that returning a copy of the Info will be good, but looks like we literally have to provide getters to ensure immutability.

Reply:
ouch. It'd have to be unexported with getter _and_ setter, because unexported fields won't be accessible from `config*.contextWithClient` since it's in another package... Info being a non-pointer also makes things difficult, so more code surface needs to change to use `*client.Info`, at least on the set-path.

Reply:
I think adding an immutable `*Metadata` fixes a lot of things, trying that route...

Reply:
Let's see, maybe just add that API without changing all the code to take a look.

Reply:
I don't think we need a mutex as map reads should be thread safe, and we return a copy of the result.